PXC-3789: PXC 8.0.27 refresh - Q4 2021 (Galera)

kamil-holubicki opened the pull request:
https://jira.percona.com/browse/PXC-3789

Merged Galera tag release_26.4.10.

Revoked fix for PXC-3024
fc1fb94965752f9163afed8fac0fb3e4d3b8cb27 later modified in
commit 4f40bea10126b65f92ebf453821656118fd92dbd) as fixed by upstream
commit e3e81d2ad789a2689e9ba330e49da228b761aedd

temeo committed:
Refactored ASIO/networking code

A wrapper interface around ASIO library was introduced and
direct calls to ASIO library were isolated under separate
compilation units under galerautils. From now on, ASIO/networking
should be used only via API defined in gu_asio.hpp.

Public API defined in gu_asio.hpp lives under gu namespace. The
following classes for client use are provided:

- AsioIpAddress for wrapping IP address representation.
- AsioConstBuffer/AsioMutableBuffer for buffer management.
- AsioSocket for stream oriented sockets and accompanied
  abstract completion handler interface AsioSocketHandler
  to support asynchronous operations.
- AsioDatagramSocket for datagram oriented messaging.
- AsioAcceptor for accepting stream oriented connections.
- AsioIoService for factory methods and running reactive
  service for sockets and timers.
- AsioSteadyTimer and associated AsioSteadyTimerHandler interface
  for timers.

For examples of API usage, see unit tests in
galerautils/tests/gu_asio_test.cpp.

On implementation side, native ASIO sockets are used in reactive manner
to detect if the socket is readable/writable. ASIO reactive
service calls completion handlers which forward the call to
stream engine object associated to AsioSocket for reading/writing the
stream. Two stream engines have implemented: One for plain TCP streams
and another for SSL/TLS streams.

A boolean CMake configuration option GALERA_WITH_SSL was introduced
to build Galera with or without SSL support. SSL related configuration
options are visible only for SSL support has been compiled in.

No changes in configuration options were introduced.

temeo committed:
Fix post-build libgalera_smm.so checks

Output of objdump contained a full path to libgalera_smm.so, which
made post-build checks to fail if the build directory path
contained string "asio".

Changed custom command to user CMAKE_BINARY_DIR as working directory
to suppress absolute path from file name argument.

temeo committed:
Merge branch '3.x-fix-post-build-checks' into 4.x-fix-post-build-checks

temeo committed:
Make sure that -DNDEBUG is defined in release builds

temeo committed:
Merge remote-tracking branch 'origin/3.x-fix-ndebug' into 4.x-fix-ndebug

temeo committed:
Compilation fixes

- Added missing `engine_` initializer in `AsioAcceptorReact`
  constructor.
- Add GU_ASIO_TEST_CERT_DIR into galerautils/tests scons CPPFLAGS.

temeo committed:
Review fixes

- Restored original behavior in `ist::Receiver::prepare()` `recv_addr_`
  rewriting.
- Cleanup commented out code.
- Removed excessive GALERA_HAVE_SSL ifdefs from gmcast. Variable
  `use_ssl` is always false if build without SSL.

Commit:
Dynamic connection engine support

Introduced new variable `socket.dyamic` which uses dynamic stream
engine. This engine is used to dynamically switch between SSL to TCP
without closing socket. Engine is using timeouts so connection have overhead.

ayurchen committed:
Fix scons build to build and run wsrep_tests

temeo committed:
codership/galera#571 Support for certificate chains

Load certificate chain file instead of certificate file in SSL context
setup.

This commit adds also certificate verification error reason to
SSL error message.

Generated certificate chains for testing purposes for three nodes
under tests/conf.

ayurchen committed:
.gitignore cleanups and stuff

ayurchen committed:
Update wsrep submodule to latest revision

ayurchen committed:
Merge branch '202101_cleanups-3.x' into 202101_cleanups-4.x

temeo committed:
Update CONTRIBUTOR_AGREEMENT.txt

temeo committed:
Merge branch '3.x-cla' into 4.x-cla

Commit:
SSL certificate dynamic reloading

Introduced global signal facility which can be used to send specific
signals to objects. Implementation of `S_CONFIG_RELOAD_CERTIFICATE`
signals which will trigger reloading of crypto context for AsioIoService
objects.

temeo committed:
codership/galera#558 Replace objdump and lld with ${CMAKE_OBJDUMP}

Use ${CMAKE_OBJDUMP} instead of build architecture objdump/lld
to allow cross compilation.

Commit:
Signed contributor agreement

temeo committed:
codership/galera#558 Check if atomic lib linkage is needed

temeo committed:
Merge branch '3.x-cmake-atomic_8' into 4.x-cmake-atomic_8

Commit:
Fix memory leak in NBO certification

Clear nbo_index_ in Certification destructor, as it may contain
entries created by transactions that did not manage to `end_nbo()`.

Fixes codership/galera-bugs#631

temeo committed:
Asio logging cleanup and wsrep-API update

- Don't print warning message on eof.
- Update wsrep-API to latest.

temeo committed:
Avoid terminate called without an active exception

GComm runner thread made pthread_exit() from member function, which
caused 'terminate called without an active exception' if there is no
other thread joining it.

Made thread to run from non-member function to avoid going out
of scope when pthread_exit() is called.

temeo committed:
Merge branch '3.x-terminate-called' into 4.x-terminate-called

ayurchen committed:
Fix wsrep_uuid_t variables alignment in galera/tests/saved_state_check.cpp

ayurchen committed:
Fix the check for "objudump" utility

ayurchen committed:
Fix compilation failure on FreeBSD due to wrong byteswap macro usage
Refs: codership/galera#581
https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=250664
https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=253612

ayurchen committed:
On FreeBSD use clang as a default compiler in scons builds

ayurchen committed:
Fix hardware CRC32C support in FreeBSD/ARM in scons builds

ayurchen committed:
Fix compilation on FreeBSD (some networking symbols)

ayurchen committed:
Fix check macros in ist_check.cpp

ayurchen committed:
Merge branch 'freebsd-bswap-3.x' into freebsd-bswap-4.x

temeo committed:
Skip post build checks on FreeBSD, fix atomics detection

Post build checks use objdump and does not work reliably
on all FreeBSD platforms. Therefore skipping for now.

Atomics library detection does not seem to work on FreeBSD,
trusting that compiler has proper support.

temeo committed:
Ported cmake patches from FreeBSD ports

temeo committed:
Merge branch '3.x-freebsd-fixes' into 4.x-freebsd-fixes

ayurchen committed:
Add CA file to galera_cert.pem and galera_key.pem pair.
Update build and test scripts

ayurchen committed:
Merge branch 'galera-ssl-fixes-3.x' into galera-ssl-fixes-4.x

temeo committed:
Fix compilation with -DGALERA_WITH_SSL:BOOL=OFF

temeo committed:
codership/galera#210 Clear grastate.dat seqno on joiner before IST

Seqno in grastate.dat was not cleared before processing IST
on joiner. As a result, if the joiner was killed during IST,
it recovered incorrect position from grastate.dat leading
to later failures.

As a fix, set seqno in grastate.dat to WSREP_SEQNO_UNDEFINED
before applying write sets from IST.

temeo committed:
Merge branch '3.x-clear-seqno-before-ist' into 4.x-clear-seqno-before-ist

Commit:
Disable DynamicStreamEngine if not compiled with SSL

ayurchen committed:
Remove #ifdef NDEBUG around try/catch clause in galera_to_execute_start() - it is
totally outdated.

Refs codership/galera-bugs#663

temeo committed:
Bump version to 25.3.33

temeo committed:
Null Merge remote-tracking branch 'origin/3.x-25.3.33' into 4.x-26.4.8

temeo committed:
Bump version number for 26.4.8

temeo committed:
Fix CRC32C detection for ARM architectures

Older compilers (GCC 4.8) don't have enough intrinsics exposed
for accelerated CRC32C on ARM. Instead of checking if the compiler
accepts the flag, check that the test code compiles.

temeo committed:
Merge branch '3.x-fix-centos7-arm' into 4.x-fix-centos7-arm

temeo committed:
codership/galera-bugs#672 Uncaught exception: Address already in use

Asio exceptions were not caught in many AsioAcceptorReact
methods, including listen().

As a fix, added try/catch blocks around methods which do
direct asio calls.

ayurchen committed:
fix incorrect assert: node in GCS_NODE_STATE_PRIM shoud be able to handle
new primary configuration. Transition PRIM-PRIM is allowed in transition
map.
Also a minor refactoring in mysql/build.sh helper script

Refs codership/galera-bugs#676

Commit:
Set depends_seqno on dummy writesets

This patch makes changes to certification so that we preserve the
depends_seqno of writesets that fail the certification test.
This change avoids the case where we deliver such a writeset to
the DBMS, which is still the process of BF aborting the same
transaction.
This may happen in the case of streaming replication, where it is
possible that the transactions manages to replicate one more fragment
before BF abort, and the fragment fails certification.

Commit:
Fixup for "Set depends_seqno on dummy writesets"

Mark writesets that fail certification as "skipped" in gcache.

Follow up for commit 0738cce2206dfe4af9707d3c33e5ee6e2e39d5de

temeo committed:
codership/galera-bugs#684 Replaying can be started too soon

Replaying transaction must wait for all the preceding
transactions to be ordered for commit to guarantee
conflict free replaying. In the original code this was
implemented by grabbing apply monitor with depends_on
set to global_seqno - 1. However, local write sets
get past apply monitor immediately regardless of the
depends_on value.

As a fix, wait for all the preceding transactions to
leave commit monitor before starting the replay.

temeo committed:
codership/galera#558 Restored CK_TIMEOUT_MULTIPLIER in debian rules

temeo committed:
Merge branch '3.x-galera-558' into 4.x-galera-558

ayurchen committed:
ANY IST failure results in state being marked "corrupt", which is too
excessive and may result in the unwarranted fallback to full SST.
During IST invalidate state only in case of trx applying failure, ignore
connection/transfer failures and such.

Refs codership/galera-bugs#681

ayurchen committed:
Merge branch 'galera-bugs#681-3.x' into galera-bugs#681-4.x

temeo committed:
codership/galera#595 Disable multicast tests by default

Multicast may not be available in all build environments,
enable only if GALERA_ASIO_TEST_MULTICAST is specified.

temeo committed:
Introduce apply_waits status variable

Count the number of times a thread enters to wait state
in apply monitor. The counter is exposed as a status variable.

ayurchen committed:
small build and test scripts cleanups

temeo committed:
Merge branch '3.x-monitor-waits' into 4.x-monitor-waits

ayurchen committed:
fix debug assert in GCache::disacard_seqno()

ayurchen committed:
if GCache::seqno_release() is called with seqno greater than the seqno
locked for IST, it goes into tight loop until the end of IST. Instead
it should just return. Fixed.

Refs codership/galera-bugs#703

ayurchen committed:
reduce the unnecessary GCache debug output verbosity.

temeo committed:
codership/galera-bugs#705 Wrong certification results after SST

Certification index was not populated after SST on joiner
for write sets which were contained in the SST but had global
seqno greater than group seqno.

As a fix, do certification for such a write sets to get
certification index populated, but do not apply the write set.

ayurchen committed:
add seesaw test script that uses sysbench for load generator

ayurchen committed:
grep away AUTO_INCREMENT values in CREATE TABLE statements for consistency
checking

ayurchen committed:
use stop_count instead of queue_len to determine if flow control is
active in gcs_replv(). This should allow lagging node (e.g. donor)
to replicate as intended if no flow contorol was activated.

Refs codership/galera-bugs#707

temeo committed:
Null Merge branch '3.x-galera-bugs-705' into 4.x-galera-bugs-705

ayurchen committed:
Log fatal error and abort instead of throwing an exception in debug
builds in key_set.cpp

ayurchen committed:
Merge branch 'gcache-issues-3.x' into gcache-issues-4.x

ayurchen committed:
handle exceptions on seqno2ptr_.insert() during GCache recovery: seqnos
there can be absolutely arbitrary and there is no guarantee that insert
will succeed.

Refs codership/galera-bugs#713

ayurchen committed:
Merge branch 'recovery-exception-3.x' into recovery-exception-4.x

temeo committed:
Bump version to 3.34

temeo committed:
Null Merge remote-tracking branch 'origin/3.x-3.34' into 4.x

temeo committed:
Bumped version numbers for 4.9

Commit:
Set `socket.ssl_reload` parameter so it is returned in provider options

ayurchen committed:
Add `local_monitor_enter_sync` point for non-transactional events
Refs codership/galera-bugs#730

ayurchen committed:
Avoid code duplication and make sure to catch exceptions in both
instances of IST sending. Fixes node stuck in the DONOR state in case
IST fails.

Refs codership/galera-bugs#679

temeo committed:
Fix static SSL library detection with CMake >= 3.3.1

- Added missing include_directories()
- Remove exact version requirement

temeo committed:
Merge branch '3.x-fix-cmake-ssl' into 4.x-fix-cmake-ssl

abychko committed:
added OpenSSL license if galera uses static SSL libs

abychko committed:
Merge branch 'galera-openssl-license' into galera-openssl-license-4.x

* galera-openssl-license:
  added OpenSSL license if galera uses static SSL libs

temeo committed:
Deprecate gcs.fc_master_slave provider option

temeo committed:
Bumped version numbers for Galera 3.35

temeo committed:
Merge remote-tracking branch 'origin/3.x-26.3.35' into 4.x-26.4.10

temeo committed:
Bumped version numbers for Galera 4.10

dutow commented:
"1. Conflicts resolved" - I do not see a separate conflict resolution commit, the commit mentioning this is the merge commit itself. What were the conflicts?

kamil-holubicki committed:
Merge tag 'release_26.4.10' into PXC-3789

# Conflicts:
#	CONTRIBUTOR_AGREEMENT.txt
#	SConstruct
#	galera/src/ist.cpp
#	galera/src/ist.hpp
#	galera/src/ist_proto.hpp
#	galera/src/replicator_str.cpp
#	galerautils/src/gu_asio.hpp
#	galerautils/src/gu_lock.hpp
#	gcomm/src/asio_tcp.cpp
#	gcomm/src/asio_udp.cpp
#	gcs/src/gcs.cpp
#	gcs/src/gcs_gcomm.cpp

kamil-holubicki committed:
Merge tag 'release_26.4.10' into PXC-3789

https://jira.percona.com/browse/PXC-3789

1. Conflicts resolved
2. Fix for PXC-3580 disable do allow compilation

kamil-holubicki committed:
PXC-3789: merge release_26.4.10

https://jira.percona.com/browse/PXC-3789

Revoked fix for PXC-3024
(commit fc1fb94965752f9163afed8fac0fb3e4d3b8cb27 later modified in
commit 4f40bea10126b65f92ebf453821656118fd92dbd) as fixed by upstream
commit e3e81d2ad789a2689e9ba330e49da228b761aedd

kamil-holubicki commented:
> "1. Conflicts resolved" - I do not see a separate conflict resolution commit, the commit mentioning this is the merge commit itself. What were the conflicts?

Fixed. It looks like I squashed 2 commits by omission.

kamil-holubicki committed:
PXC-3789: merge release_26.4.10

https://jira.percona.com/browse/PXC-3789

wsrep-API submodule pointer updated
